### PR TITLE
Support redirects.yaml.

### LINKED ISSUE #11085: Feature: redirects.yaml
Our redirect feature is cool and powerful, but having to correctly format a JSON file by hand is HORRIBLE.

We are now using YAML for user-edited settings files, we should do that for redirects as well.

## Ref material

Our current JSON redirects format looks like:

```
{
  "from": "/url/" ,
  "to":  "/url/",
  "permanent": true | false
}
```

- We support full JavaScript regexes in the `from` field
- Regex patterns e.g. $1, $2 so on can be substituted into the `to` field.
- Redirects can be either permanent (301) redirects, or temporary (302) redirects, based on the permanent flag.
- Also (currently undocumented) we support adding an `i` flag to the end of the `from` path, e.g. `"from": "/my-url/i"` which makes it do a case insensitive match.

Statamic have a similar feature, provided in YAML: https://docs.statamic.com/routing#redirects

Netlify use a tab separated values style: https://www.netlify.com/docs/redirects/

## Spec

We will add support for uploading a redirects file called `redirects.yaml`. The file must contain valid YAML.  

Redirects will apply to the web application of Ghost only (same as now).  

This new file will support all the same features as the old redirects.json format but in an easier to write and maintain style. The main change is that instead of a property, the type of redirect is specified via one of two top-level keys.

Full example:

```
302:
  /from-url/: /to-url/

301:
  /category/([a-z0-9\-]+)/i: /tag/$1/
  /v([0-9\.]+)/docs/([a-z0-9\-]+)/i: /docs/$2/
```

## Migration

For now we should just support both formats. In the next major version we can drop support for uploading the JSON format. 

We may be able to write a migration for a major upgrade using something like https://www.npmjs.com/package/json2yaml

However JSON is valid YAML 🤔

### PR DESCRIPTION
**Pull request by sainthkh**
Closes #11085 

Thanks @assisting!

## Todos

- [x] Allow current Admin UI to accept both JSON and YAML file. => TryGhost/Ghost-Admin#1703
- [x] Make server save the redirects file in yaml format. 
- [x] Let users download the current redirects file. (Extension doesn't matter.)
- [x] Add related tests.
- [x] Add a test about invalid yaml file.

----

- [x] There's a clear use-case for this code change
- [x] Commit message has a short title & references relevant issues
- [x] The build will pass (run `yarn test` and `yarn lint`)

### CODE REVIEW

File: core/server/web/shared/middlewares/custom-redirects.js
Comment (naz) [inline core/server/web/shared/middlewares/custom-redirects.js:30]: Nice, think this is a right direction with code extraction :+1:

File: core/server/web/shared/middlewares/custom-redirects.js
Comment (naz) [inline core/server/web/shared/middlewares/custom-redirects.js:32]: Maybe instead of giving the `parseRedirectsFile` two parameters we could give it just the `filepath` and let the logic detecting the extension live inside of the `parseRedirectsFile` method?

File: core/server/api/canary/redirects.js
Comment (naz) [inline core/server/api/canary/redirects.js]: What would be a response with this `plan` format look like? Assuming it adjusts to whatever the `redirects` file type is? (don't remember from the top of my head)

File: core/server/api/canary/redirects.js
Comment (naz) [inline core/server/api/canary/redirects.js:17]: :ok_hand: good call!

File: core/frontend/services/redirects/settings.js
Comment (naz) [inline core/frontend/services/redirects/settings.js]: Would rather name `defaultJsonFileContent` as `defaultRedirectsContent`

File: core/frontend/services/redirects/settings.js
Comment (naz) [inline core/frontend/services/redirects/settings.js:175]: I'm torn between handling this situation hear vs leaving it to [readRedirectsFile](https://github.com/TryGhost/Ghost/pull/12187/commits/b6252efe576c0e766f7987231f393e04ef901853#diff-48644be82a9b957e5e627bf7b0f2f73cdb1d63851ffad68c7c178c5886495bb8R16) to return the `[]` as it would in case the file does not exist. What are your thoughts?

File: core/server/api/canary/redirects.js
Comment (naz) [inline core/server/api/canary/redirects.js:29]: Hm... not really convinced about having `sync` method here... I know it's necessary in this situation, but would block the even loop. There could be couple solutions here:
1. Add support for async `response.format()` handling, then we'd be able to return an promise here and use async method to figure out redirects file path.
2. "cache" file path inside of `redirects.settings` module and return cached value here

File: test/regression/api/canary/admin/redirects_spec.js
Comment (naz) [inline test/regression/api/canary/admin/redirects_spec.js:295]: We'd probably want to change the endpoint name to something like `redirects/upload` and `redirects/download` instead of keeping the format name here. That would be similar to the naming use for members upload/download endpoints -  https://github.com/TryGhost/Ghost/blob/635580f291bd826c9119049ad968cb7702477ba7/core/server/web/api/canary/admin/routes.js#L95-L96

File: core/server/web/api/canary/admin/routes.js
Comment (naz) [inline core/server/web/api/canary/admin/routes.js:223]: This is a `canary` route and it hasn't been marked as "stable" anywhere iirc. Would need to give a thought about renaming this endpoint in current version :thinking:

File: core/frontend/services/redirects/settings.js
Comment (naz) [inline core/frontend/services/redirects/settings.js:131]: Adding a comment about why this default is here would be good. Specifically that API v2 used to support only `.json` extension and this default should go away with `Ghost v4` :)

File: core/server/translations/en.json
Comment (naz) [inline core/server/translations/en.json:361]: @peterzimon this needs a copy review from you. The context of the error is available here - https://github.com/TryGhost/Ghost/pull/12187/commits/9ddfdcdecdfeb5e75273f5f87a2ec460a36cf3fe#diff-48644be82a9b957e5e627bf7b0f2f73cdb1d63851ffad68c7c178c5886495bb8R48-R51. It might happen when user upload a `yaml` file which is just a "string" without any specific data (incorrect format).

**Review by naz — CHANGES_REQUESTED** [review]
This is an awesome contribution @sainthkh ! Thanks a lot. I've added a few comments and change requests. Will be able to come back to this in another 2 weeks time :crossed_fingers:

**Comment by sainthkh** [comment]
@naz I've read your review. I'm currently focusing on #12180. Because of that, my response to your reviews would take a bit longer than expected.

File: core/server/api/canary/redirects.js
Comment (sainthkh) [reply core/server/api/canary/redirects.js]: By default, Ghost server responds in `json` format. 

https://github.com/TryGhost/Ghost/blob/43b7c7b3444aa695a5cc044c65e065ac0262c9ab/core/server/api/shared/http.js#L93-L94

But if we use this format with `yaml` file, the downloaded file is converted to JSON string. It's not what we want. That's why we need to use `plain` text for `yaml` files.

File: core/frontend/services/redirects/settings.js
Comment (sainthkh) [reply core/frontend/services/redirects/settings.js]: I think `defaultJsonFileContent` is the right name. Because json redirects file is the default until it is deprecated. 

The current default content for json file is the `[]`.

Comment (sainthkh) [reply core/frontend/services/redirects/settings.js]: I reviewed it today and your name sounds better. I'll change it that way.

Comment (sainthkh) [reply core/frontend/services/redirects/settings.js]: And the default file content for 'yaml' file is the empty string, not `[]`.

File: core/frontend/services/redirects/settings.js
Comment (sainthkh) [reply core/frontend/services/redirects/settings.js:175]: I checked the code and decided not to change things because `readRedirectsFile` is also used at `setFromFilePath`. If we add it to `readRedirectsFile`, it causes error because `JSON.parse([])` throws an error. 

I also thought about returning string `'[]'` inside `readRedirectsFile`. But it is also not a good idea because it's a bit unintuitive. It's not the result of reading a file.

File: core/server/api/canary/redirects.js
Comment (sainthkh) [reply core/server/api/canary/redirects.js:29]: I tried the cache method. But I decided not to use it because users can change json to yaml or vice versa with other tools like ftp, telnet. And it makes testing almost impossible. 

That's why I implemented the 1.

File: core/frontend/services/redirects/settings.js
Comment (sainthkh) [reply core/frontend/services/redirects/settings.js:131]: I've left a comment about that.

File: core/server/web/shared/middlewares/custom-redirects.js
Comment (sainthkh) [reply core/server/web/shared/middlewares/custom-redirects.js]: I've created a `loadRedirectsFile` method and removed `filePath` and `content`.

File: core/frontend/services/redirects/settings.js
Comment (sainthkh) [reply core/frontend/services/redirects/settings.js:3]: OK. I restored it.

File: test/regression/api/canary/admin/redirects_spec.js
Comment (sainthkh) [reply test/regression/api/canary/admin/redirects_spec.js:295]: Closing this in favor of the comment below.

File: core/server/web/api/canary/admin/routes.js
Comment (sainthkh) [reply core/server/web/api/canary/admin/routes.js:223]: I love your idea of naming them `redirects/upload`, `redirects/download`. They're more intuitive. 

But I'm a bit against removing current `redirects/json` APIs. Because it's a breaking change and it's against the semver rule.

My suggestion is to just add `redirects/upload`, `redirects/download` and remove `redirect/json` in v4.

File: core/server/api/canary/redirects.js
Comment (naz) [reply core/server/api/canary/redirects.js]: If someone is changing the file directly we usually assume the user knows what they are doing and would restart the instance manually. For example, if you modify `routes.yaml` file right now through ftp these changes won't kick in till restart. Cache should be refreshed when the user uploads a new file through API, that's the only way we can control it and should be enough.

File: core/server/web/api/canary/admin/routes.js
Comment (naz) [reply core/server/web/api/canary/admin/routes.js:223]: Think your suggestion is the best way forward :+1: We'd need to switch over Ghost-Admin to use the new endpoint once these changes are merged. Can you add a note in the main issue or create a Ghost-Admin PR linking this this issue please? This way both changes can go into master together.

File: core/server/api/canary/redirects.js
Comment (naz) [inline core/server/api/canary/redirects.js:45]: Don't think we need this parameter here. We are passing file path already, so inside of `setFromFilePath` method we can get the extension through `const {ext} = path.parse(filePath);`

File: core/server/api/shared/http.js
Comment (naz) [inline core/server/api/shared/http.js:88]: Super nice :+1:

File: core/server/api/canary/redirects.js
Comment (naz) [reply core/server/api/canary/redirects.js]: For now, I think the async method will be enough and we can think about caching the response if it becomes problematic. Sites rarely need to change these redirects after initial setup.

**Comment by naz** [comment]
@sainthkh I think these changes are ready to go into master. There are couple "nice to have" changes: adding [aliased routes](https://github.com/TryGhost/Ghost/pull/12187#discussion_r511699113) and [changing](https://github.com/TryGhost/Ghost/pull/12187#discussion_r515717389) `setFromFilePath` method signature. Also, need a [copy review ](https://github.com/TryGhost/Ghost/pull/12187#discussion_r508144220) from @peterzimon. Think I'll merge these changes "as is" after the copy review.  

Wanted to thank you for making this feature production ready. Awesome work!

**Review by naz — APPROVED** [review]

File: core/server/api/canary/redirects.js
Comment (sainthkh) [reply core/server/api/canary/redirects.js:45]: As for this, I'll check this out tomorrow. If I remember correctly, the `frame.file.path` doesn't have an extension in Windows.

File: core/server/translations/en.json
Comment (naz) [reply core/server/translations/en.json:373]: @justin-ghost ping on the above

Comment (justin-ghost) [reply core/server/translations/en.json:373]: @naz `invalidFile` copy looks good. For `yamlParse`, here are two options to consider based on the ambiguity of the error: 

* YAML output cannot be a plain string. Check the format of your YAML file. 
* Incorrect file format. Check the format of your YAML file.

Comment (sainthkh) [reply core/server/translations/en.json:373]: @justin-ghost I prefer the first one because it tells users what went wrong.

File: core/server/api/canary/redirects.js
Comment (sainthkh) [reply core/server/api/canary/redirects.js:45]: image

I printed out the `frame.file` and it became clear that we cannot use `frame.file.path` because there is no extension.

Comment (naz) [reply core/server/api/canary/redirects.js:45]: Cool. Let's not change this for now than. Good to know where this method signature comes from :+1:

File: core/server/translations/en.json
Comment (naz) [reply core/server/translations/en.json]: Nice. Good call with 1st option @sainthkh . Thanks for feedback @justin-ghost :black_heart:

Comment (naz) [reply core/server/translations/en.json]: After having a little more thought think the message is meant to say "YAML ~output~ **input**  cannot be a plain string". We are dealing with an **input** file here. @justin-ghost let me know if this makes sens. I will go on and merge the changes no, we could modify the copy a little in master after

**Comment by naz** [comment]
It's in master :tada: Great job @sainthkh :fist_raised:

**Pull request merged**
